packages/ses/types.d.ts

```diff
@@ -59,7 +59,7 @@ export interface ThirdPartyStaticModuleInterface {
   imports: Array<string>;
   exports: Array<string>;
   execute(
-    proxiedExports: Object,
+    proxiedExports: Record<PropertyKey, any>,
```

Comment:
I'll defer the review of this change to @kriskowal

Reply:
The problem with `Object` is that you cannot assign to it.  You can't assign to `object`, either.

Reply:
And I defer to the submitter since we can rely on @boneskull to answer support questions on Matrix going forward :-). I have no idea. Did not know `PropertyKey` was a TS built-in.

Reply:
Yep! It's `string | number | symbol` from the `es5` internal lib.

Reply:
Is this object parameter meant to be mutated? That seem contrary to our programming model of hardening what is passed.

Reply:
The `proxiedExports` is the internal view of the namespace and `exportsProxy` is the external view of the namespace. The internal view is mutated by bindings during evaluation of the module functor, and the external view needs to be consistent with namespace objects, which are not mutable.

The names are bad. Since I wrote this, I learned about “module environment record” which is approximately the 262 name for the internal object I’m calling `proxiedExports`. Not precisely. And it’s an awful name too.

Reply:
that's all over my head, but this is what prompted the change:

```js
export const importHook = async (specifier) => {
  const ns = await import(specifier)
  return Object.freeze(
    /** @type {import('ses').ThirdPartyStaticModuleInterface} */ ({
      imports: [],
      exports: Object.keys(ns),
      execute: (moduleExports) => {
        moduleExports.default = ns // <--- TYPE ERROR
        Object.assign(moduleExports, ns)
      },
    })
  )
}
```

Reply:
@kriskowal Also, it shouldn't be breaking anybody to change the name of that argument.  I can change it to something else..?

Reply:
Let’s reserve renaming to another change, which you’re welcome to propose! I am dissatisfied with the name but don’t have a better idea yet. Ideas welcome! `namespace` is a good name for `proxiedExports` if there’s a dual that doesn’t have `namespace` in its name. `privateNamespace` vs `publicNamespace` or `internalNamespace` vs `externalNamespace` would be clearer albeit tortured. I like `bindings` except that I want to use that word for a property of a `ModuleSource` object that reflects the “bindings” between the internal and external namespace objects.

Reply:
I should note that `bindings` is the name that our friends at Moddable suggested and it was such a good name that it made clear in my mind how these two objects are related. The `as` keyword does a lot of work in a “binding” like `export { name as rename }`. That would create a “binding” between `internal.name` and `external.rename` for a pair of module namespaces. It would be reasonable to call both namespaces and distinguish them by context in most places.

I might buy (external) `namespace` and (internal) `environment`.

Reply:
Yeah if the argument is meant to be assigned to, `Record` is the right type. It is just a surprising API shape.

Reply:
Oh, also worth noting that `PropertyKey` might be too broad. I don’t think modules can import or export symbol-named or nominal properties. I _do_ know that the strings aren’t not limited to valid JavaScript identifiers!

Reply:
> Oh, also worth noting that PropertyKey might be too broad. I don’t think modules can import or export symbol-named or nominal properties. I do know that the strings aren’t limited to valid JavaScript identifiers!

That appears to be true.  It's not true about CJS, of course.  If we narrow the type further to `Record<string, any>`, this _could_ impact something working with a CJS package, though perhaps unlikely.

How about this: let's do `Record<string, any>` for now, and if it becomes a problem, we change it to `Record<PropertyKey, any>` later. That won't be a breaking change, but narrowing the type arguably would.

Reply:
I've updated the code to do that.

Reply:
Summoning @naugtur to rule on the ESM<->CJS namespace bindings behavior for symbols. My perception is that they would appear on `namespace.default` but not be replicated as `namespace[symbol]`.

Reply:
I'm gonna need to look into it to make sure, but off the top of my head I think there's no way to define a symbol key in the record's exports but on the other hand my implementation of lifting them from defaults is not explicitly skipping symbols. Gotta dig deeper.

Reply:
@naugtur Understood.  This begs the question, though, _should_ `Object.keys()` be used there, and not, say, `Object.getOwnPropertyDescriptors()`?

Reply:
I've updated the comment.

Reply:
I'm not sure if it's possible for whatever Object.keys would skip to be detected as named exports by the lexer and only the pre-defined top-level fields will be able to populate from cjs to esm. But I can't say you're not right pointing that out.

One case I can think of is having a getter for lazy loading. I wonder if copying a getter over would even work as intended.